docs/dev/testing.md

```diff
@@ -90,32 +99,162 @@ reject file:
 change before overwriting `.result` files.
 
+### Debugging CQL tests
+
+To debug CQL tests, one can run CQL against a standalone
+scylla (possibly started in debugger) using `cqlsh`.
+
 ## Unit tests
 
 The same unit test can be run in different seastar configurations, i.e. with
 different command line arguments. The custom arguments can be set in
 `custom_args` key of the `suite.yaml` file.
 
-
-## Debugging
+### Debugging unit tests
 
 If a test fails, its log can be found in `testlog/${mode}/testname.log`.
 By default, all unit tests are built stripped. To build non-stripped tests,
 `./configure` with `--tests-debuginfo list-of-tests`.
 `test.py` adds some command line arguments to unit tests. The exact way in
 which the test is invoked is recorded in `testlog/test.py.log`.
 
-To debug CQL tests, one can invoke `gdb cql_repl`, which will start
-`cql_repl` in interactive mode, and then supply faulty CQL statements
-on the terminal.
-
-
-## Automation
+## Python tests
+
+`test.py` supports pytest standard of tests, for suites (directories)
+specifying 'Python' test type in their scylla.yaml. For such tests,
+a standalone server instance is created, and a connection URI to the
+server is passed to the test. Thanks to conveninence fixtures,
+test writers don't need to create or cleanup connections or keyspaces.
+`test.py` will also keep track of the used server(s) and will shut
+down the server when all tests  using it end.
+
+## Sharing and pooling servers
+
+Since there can be many pytests in a single directory (e.g. cql-pytest)
+`test.py` creates multiple servers to parallelize their execution.
+Each server is also shared among many tests, to save on setup/teardown
+steps. While this speeds up execution, sharing servers complicates debugging
+if a test fails.
+
+Specifically, you should avoid leaving global artifacts in your test, even
+if it fails. Typically, you could use a built-in `keyspace()` fixture
+to create a randomly named keyspace.
+
+At start and end of each test, `test.py` performs a number of sanity checks
+of the used server:
+- it should be up and running,
+- it should not contain non-system keyspaces.
+
+### Debugging a pytest.
+
+To have a full picture for a failing pytest it is necessary to identify
+the server which was used to run it and the relevant framgnet in the server
+log. Thankfuly, `test.py` maintains this link through relevant log
+messages and preserves Scylla output on test failure.
+
+A typical debugging journey should start with looking at `test.py.log` in
+`testlog` where, for each test it runs, `test.py` prints all relevant paths
+to server log and pytest output.
+
+To extend `test.py` logging, you can use the standard 'logging' module API.
+Individual pytests are programmed to not gobble stdout, so you can can also
+add prints to pytests, and they will end up in the test' log.
+
+For example, imaigine `cql-pytest/test_null.py` fails. The relevant lines
+in test.py.log will be:
```

Comment:
test.py in single quotes for consistency?

Reply:
fixed